### ISSUE TITLE
concat and BN problem for inputting image data and vector data simultaneously

### ISSUE THREAD

**Issue by yellowpuppy** [BVLC/caffe#6362]
I have a labeled RGB image dataset and a external vector data as input,  the vector data and image data is corresponding, and finally I use  L2-Loss to regress the label value.  I use AlexNet for image and two FC layer for vector input, and I concat two branch before FC6 in AlexNet (the second FC layer in AlexNet with 4096 neuron) ,THE PROBLEM IS, after concat, the performance did not improve comparing with only input image data using AlexNet (the vector data turly related to the label). Moreover, when I add BN and Scale layer before each ReLU in the concated NET, the TEST loss decrease at first ,but after a little while then increase. However, when I discard BN，the test loss is OK, why?    Is the structure problem?  How to design it?  Thanks!

`name: "AlexNet+mynet-BN"
layer {
  name: "data"
  type: "HDF5Data"
  top: "data1"
  top: "data2"
  top: "score"
  include {
    phase: TRAIN
  }
  hdf5_data_param {
    source: "/media/yq/yq2/CNN/myfile24/train_all_N_twoset_h5.txt"
    batch_size: 64
    shuffle: true
  }

}
layer {
  name: "data"
  type: "HDF5Data"
  top: "data1"
  top: "data2"
  top: "score"
  include {
    phase: TEST
  }

  hdf5_data_param {
    source: "/media/yq/yq2/CNN/myfile24/test_total_N_twoset_h5.txt"
    batch_size: 256
    shuffle: true
  }
}
#---------------------------------------------data2----------------------------
layer {
  name: "bn0/data2"
  type: "BatchNorm"
  bottom: "data2"
  top: "data2/bn"
  
}
layer {
  name: "scale0/data2"
  type: "Scale"
  bottom: "data2/bn"
  top: "data2/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "fc1/data2"
  type: "InnerProduct"
  bottom: "data2/scale"
  top: "fc1/data2"
  param {
    lr_mult: 1
    decay_mult:  1                             #change
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 512
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn1/data2"
  type: "BatchNorm"
  bottom: "fc1/data2"
  top: "fc1/data2/bn"
  
}
layer {
  name: "scale1/data2"
  type: "Scale"
  bottom: "fc1/data2/bn"
  top: "fc1/data2/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu1/data2"
  type: "ReLU"
  bottom: "fc1/data2/scale"
  top: "fc1/data2/scale"
}

layer {
  name: "drop1/data2"
  type: "Dropout"
  bottom: "fc1/data2/scale"
  top: "fc1/data2/scale"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "fc2/data2"
  type: "InnerProduct"
  bottom: "fc1/data2/scale"
  top: "fc2/data2"
  param {
    lr_mult: 1
    decay_mult:  1                             #change
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 1024
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn2/data2"
  type: "BatchNorm"
  bottom: "fc2/data2"
  top: "fc2/data2/bn"
  
}
layer {
  name: "scale2/data2"
  type: "Scale"
  bottom: "fc2/data2/bn"
  top: "fc2/data2/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu2/data2"
  type: "ReLU"
  bottom: "fc2/data2/scale"
  top: "fc2/data2/scale"
}

#---------------------------------------------data1----------------------------
layer {
  name: "bn0"
  type: "BatchNorm"
  bottom: "data1"
  top: "data1/bn"
  
}
layer {
  name: "scale0"
  type: "Scale"
  bottom: "data1/bn"
  top: "data1/sc"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "data1/sc"
  top: "conv1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 96
    kernel_size: 11
    stride: 4
    pad: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn1"
  type: "BatchNorm"
  bottom: "conv1"
  top: "conv1/bn"
  
}
layer {
  name: "scale1"
  type: "Scale"
  bottom: "conv1/bn"
  top: "conv1/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu1"
  type: "ReLU"
  bottom: "conv1/scale"
  top: "conv1/scale"
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "conv1/scale"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "conv2"
  type: "Convolution"
  bottom: "pool1"
  top: "conv2"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  convolution_param {
    num_output: 256
    pad: 2
    kernel_size: 5
    #group: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn2"
  type: "BatchNorm"
  bottom: "conv2"
  top: "conv2/bn"
  
}
layer {
  name: "scale2"
  type: "Scale"
  bottom: "conv2/bn"
  top: "conv2/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu2"
  type: "ReLU"
  bottom: "conv2/scale"
  top: "conv2/scale"
}
layer {
  name: "pool2"
  type: "Pooling"
  bottom: "conv2/scale"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "conv3"
  type: "Convolution"
  bottom: "pool2"
  top: "conv3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  convolution_param {
    num_output: 384
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn3"
  type: "BatchNorm"
  bottom: "conv3"
  top: "conv3/bn"
  
}
layer {
  name: "scale3"
  type: "Scale"
  bottom: "conv3/bn"
  top: "conv3/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu3"
  type: "ReLU"
  bottom: "conv3/scale"
  top: "conv3/scale"
}
layer {
  name: "conv4"
  type: "Convolution"
  bottom: "conv3/scale"
  top: "conv4"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  convolution_param {
    num_output: 384
    pad: 1
    kernel_size: 3
    #group: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn4"
  type: "BatchNorm"
  bottom: "conv4"
  top: "conv4/bn"
  
}
layer {
  name: "scale4"
  type: "Scale"
  bottom: "conv4/bn"
  top: "conv4/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu4"
  type: "ReLU"
  bottom: "conv4/scale"
  top: "conv4/scale"
}
layer {
  name: "conv5"
  type: "Convolution"
  bottom: "conv4/scale"
  top: "conv5"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  convolution_param {
    num_output: 256
    pad: 1
    kernel_size: 3
    #group: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn5"
  type: "BatchNorm"
  bottom: "conv5"
  top: "conv5/bn"
  
}
layer {
  name: "scale5"
  type: "Scale"
  bottom: "conv5/bn"
  top: "conv5/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu5"
  type: "ReLU"
  bottom: "conv5/scale"
  top: "conv5/scale"
}
layer {
  name: "pool5"
  type: "Pooling"
  bottom: "conv5/scale"
  top: "pool5"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}

layer {
  name: "fc6"
  type: "InnerProduct"
  bottom: "pool5"
  top: "fc6"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn6"
  type: "BatchNorm"
  bottom: "fc6"
  top: "fc6/bn"
  
}
layer {
  name: "scale6"
  type: "Scale"
  bottom: "fc6/bn"
  top: "fc6/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu6"
  type: "ReLU"
  bottom: "fc6/scale"
  top: "fc6/scale"
}

#---------------------------------------------------add---------------------------

layer {
  name: "concat1"
  type: "Concat"
  bottom: "fc6/scale"
  bottom:"fc2/data2/scale"
  top: "concat1"
}
layer {
  name: "drop6"
  type: "Dropout"
  bottom: "concat1"
  top: "concat1"
  dropout_param {
    dropout_ratio: 0.5
  }
}
#---------------------------------------------------add---------------------------
layer {
  name: "fc7"
  type: "InnerProduct"
  bottom: "concat1"
  top: "fc7"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn7"
  type: "BatchNorm"
  bottom: "fc7"
  top: "fc7/bn"
  
}
layer {
  name: "scale7"
  type: "Scale"
  bottom: "fc7/bn"
  top: "fc7/scale"
  scale_param {
    bias_term: true
  }
}
layer {
  name: "relu7"
  type: "ReLU"
  bottom: "fc7/scale"
  top: "fc7/scale"
}
layer {
  name: "drop7"
  type: "Dropout"
  bottom: "fc7/scale"
  top: "fc7/scale"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "fc8"
  type: "InnerProduct"
  bottom: "fc7/scale"
  top: "fc8"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }

  inner_product_param {
    num_output: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "bn8"
  type: "BatchNorm"
  bottom: "fc8"
  top: "fc8/bn"
  
}
layer {
  name: "scale8"
  type: "Scale"
  bottom: "fc8/bn"
  top: "fc8/scale"
  scale_param {
    bias_term: true
  }
}
layer {                         #change
  name: "Sigmoid8"
  type: "Sigmoid"
  bottom: "fc8/scale"
  top: "pred"
}
layer {
  name: "loss"
  type: "EuclideanLoss"
  bottom: "pred"
  bottom: "score"
  top: "loss"
}
`

**Comment by Noiredd** [comment]
Please post usage, installation, or modeling questions, or other requests for help to the [caffe-users list][1] instead of Issues. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe. Please read the [guidelines for contributing][2] before submitting an issue or a pull request.

For improved readability, please attach long prototxt instead of pasting them directly into your post.

 [1]: https://groups.google.com/forum/#!forum/caffe-users
 [2]: https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md
